### PR TITLE
gcc: add callFile to make it more convenient to call files

### PR DESCRIPTION
**Pull request by Artturin**
there is a duplication issue in the gcc's so this will help with that

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by amjoseph-nixpkgs** [comment]
This is awesome, but we need to be really really careful not to trip over https://github.com/NixOS/nixpkgs/issues/216083 -- there are a lot of calls here where both the `callPackage` site and the package being called have a default argument, and nixpkgs does the opposite of what is most obvious when that happens.

If there is a way to invert the priority (callsite supersedes callee), I think `callFile` should use the inverted priority.  I will investigate that.

**Comment by amjoseph-nixpkgs** [comment]
#215156 is totally awesome, thank you so much.

Regarding this PR, I totally agree in principle, but right now if people forget to update these `jq`-generated lists (like [I did](https://github.com/NixOS/nixpkgs/pull/209870/commits/b5c46c97c9ca53fa8ef7d123ab6b28d39fd0a68e) [twice](https://github.com/NixOS/nixpkgs/pull/209870/commits/7aacac89c3a0248a7a74a8827a2dd4d8b5c6a9a8)) there is no error message and the failure mode is really hard to troubleshoot due to the multiple layers of default arguments.

Is there any chance I could talk you into reverting this, at least temporarily?  It's new enough (less than 24 hours old) that it will revert cleanly right now.  But that won't be the case for much longer.

After looking at how Nix handles default arguments, and Tvix's reconstruction of the rules, I'm beginning to think that the only sane way to get the automatic behavior we are after here is to use `lib.newScope` like `pkgs/top-level/all-packages.nix` does.

**I volunteer to implement this,** but:

1. I am super-backlogged right now for at least the next week, and probably more like two weeks
2. I need to prioritize #209870 until it reaches some sort of conclusion

**Comment by Artturin** [comment]
There is no lib.newScope  there is pkgs.newScope which does the same thing I'm doing in this pr https://github.com/NixOS/nixpkgs/blob/14cb2130bb78b96853e576d21ea823b1b19f39de/pkgs/top-level/splice.nix#L147

> but right now if people forget to update these jq-generated lists (like I did twice) there is no error message and the failure mode is really hard to troubleshoot due to the multiple layers of default arguments.

This isn't a regression, instead it's an improvement because previously you had to pass the arguments manually to all the imports

**Comment by amjoseph-nixpkgs** [comment]
> There is no lib.newScope

Sorry I meant `lib.makeScope`.

It is powerful but badly under-documented.  [Here is an example](https://git.sr.ht/~amjoseph/ownerboot/tree/master/item/src/default.nix#L9) and [some related prose](https://git.sr.ht/~amjoseph/ownerboot/tree/master/doc/architecture.md#nixpkgs).

> because previously you had to pass the arguments manually to all the imports

Yes, it's definitely an improvement.  You do still have to list them manually (all the manual lists have been unioned together).  I think `lib.makeScope` would let us eliminate the list entirely.  When I have a working example I'll send a PR.

File: pkgs/development/compilers/gcc/common/configure-flags.nix [hunk 1]
```diff
@@ -25,7 +25,6 @@
 , langJit
 }:
 
-assert cloog != null -> lib.versionOlder version "5";
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/gcc/common/configure-flags.nix:28]: Always happy to see those asserts being removed.

**Comment by SuperSandro2000** [comment]
I just had the idea to do it like this, not to use lib.makeScope because I actually know very little around scope's in nixpkgs.

**Pull request merged**